README.adoc
= Data4Life Crypto SDK
:link-repository: https://github.com/d4l-data4life/hc-crypto-sdk-kmp
:project-version: 1.16.1
:doctype: article
:!showtitle:
:toc: macro
:toclevels: 2
:toc-title:
:icons: font
:imagesdir: assets/images
ifdef::env-github[]
:warning-caption: :warning:
:caution-caption: :fire:
:important-caption: :exclamation:
:note-caption: :paperclip:
:tip-caption: :bulb:
endif::[]

++++
<div align="center">
    <p><!-- PROJECT LOGO -->
        <a href="https://github.com/d4l-data4life/hc-crypto-sdk-kmp">
            <img src="assets/images/d4l-logo.svg" width="150"/>
        </a>
    </p>
    <p><!-- PROJECT TITLE -->
        <h3>Data4Life Crypto SDK</h3>
    </p>
    <p><!-- PROJECT DESCRIPTION -->
        Implementation of the D4L Crypto Protocol
    </p>
    <p><!-- PROJECT DOCUMENTATION -->
        <a href="README.adoc"><strong>Explore the docs »</strong></a>
    </p>
    <p><!-- PROJECT ISSUES/FEATURES -->
        <a href="https://github.com/d4l-data4life/hc-crypto-sdk-kmp/issues">Report Bug</a>
        ·
        <a href="https://github.com/d4l-data4life/hc-crypto-sdk-kmp/issues">Request Feature</a>
    </p>
    <p><!-- PROJECT BADGES see badges.adoc how to change them -->
        <a href="https://github.com/d4l-data4life/hc-crypto-sdk-kmp/releases">
            <img src="assets/images/badge-release-latest.svg" alt="Latest release"/>
        </a>
        <a href="https://github.com/d4l-data4life/hc-crypto-sdk-kmp/actions">
            <img src="https://github.com/d4l-data4life/hc-crypto-sdk-kmp/actions/workflows/d4l-ci-latest-version.yml/badge.svg" alt="D4L CI - Latest Version"/>
        </a>
        <a>
            <img src="assets/images/badge-crypto-support.svg" alt="Crypto"/>
        </a>
        <a>
            <img src="assets/images/badge-platform-support.svg" alt="Platform"/>
        </a>
        <a href="LICENSE">
            <img src="assets/images/badge-license.svg" alt="License"/>
        </a>
    </p>
</div>
++++

[discrete]
=== Table Of Contents
toc::[]

== About The Project

The Data4Life Crypto SDK is an implementation of the D4L crypto protocol in version 1.

It supports encryption and decryption of data that should be stored in the Personal Health Data Platform (PHDP).

=== Features

* Crypto protocol 1

=== Built With

* link:https://kotlinlang.org/docs/reference/mpp-intro.html[Kotlin Multiplatform]
* link:https://www.bouncycastle.org/[Bouncy Castle]

== Getting Started

Instructions how to get software up and running.

=== Prerequisites

* link:https://developer.android.com/studio#downloads[Android Studio 7.2.2]
* link:https://adoptopenjdk.net/[Java 11] | link:https://developer.android.com/studio/write/java8-support[Limitations of Java language features and APIs] | https://jakewharton.com/d8-library-desugaring/[Desugaring]
* link:https://kotlinlang.org/[Kotlin 1.7.10]
* Android 6.0 (API 23) to Android 13.0 (API 33)

=== Installation

A step by step series of examples that tell you how to get it running.

==== Consume GitHub Packages

We use GitHub Packages to distribute the SDK. In order to consume our dependencies you need to generate a GitHub Personal Access Token. Please follow the how to link:https://docs.github.com/en/packages/learn-github-packages/introduction-to-github-packages#authenticating-to-github-packages[authenticate to GitHub Packages, window="_blank"].

NOTICE:: You need to have _read_ access to this repository and generate a https://github.com/settings/tokens/new/[personal access token] with `repo` and `read:packages` scope.

===== Token

The token needs to be made available.

. Add `gpr.user = {GitHub username}` and `gpr.key = {GitHub Personal Access Token}` to your global Gradle properties `~/.gradle/gradle.properties`

    gpr.user=github-username
    gpr.key=github-token

. Or add following environment variables `PACKAGE_REGISTRY_USERNAME={GitHub username}` and `PACKAGE_REGISTRY_TOKEN={GitHub Personal Access Token}`

===== Setup Maven Repository

Add the following maven repository configuration to your root build.gradle:

[source, Gradle]
----
    allprojects {
        repositories {
            ...
            maven {
                url = URI("https://maven.pkg.github.com/d4l-data4life/hc-crypto-sdk-kmp")
                credentials {
                    username = project.findProperty("gpr.user") as String? ?: System.getenv("PACKAGE_REGISTRY_USERNAME")
                    password = project.findProperty("gpr.key") as String? ?: System.getenv("PACKAGE_REGISTRY_TOKEN")
                }
            }
        }
    }
----

==== Java

Add following dependencies to your app build.gradle.

----
dependencies {
    implementation "care.data4life.hc-crypto-sdk-kmp:crypto-jvm:$version"
}
----

==== Android

----
dependencies {
    implementation "care.data4life.hc-crypto-sdk-kmp:crypto-android:$version"
}
----

== Usage

TODO

== Roadmap

This project is work in progress. We are working on adding more functionality, guidelines, documentation and other improvements.

Next planed features:

* move more crypto code from HC-SDK-KMP

Also see the open link:{link-repository}/issues[issues] for a list of proposed features and known issues.

== Changelog

See link:CHANGELOG.adoc[changelog]

== Versioning

We use http://semver.org/[Semantic Versioning] as a guideline for our versioning.

Releases use this format: `{major}.{minor}.{patch}`

* Breaking changes bump `{major}` and reset `{minor}` & `{patch}`
* Backward compatible changes bump `{minor}` and reset `{patch}`
* Bug fixes bump `{patch}`

== Contributing

You want to help or share a proposal? You have a specific problem? Read the following:

* link:CODE-OF-CONDUCT.adoc[Code of conduct] for details on our code of conduct.
* link:CONTRIBUTING.adoc[Contributing] for details about how to report bugs and propose features.
* link:DEVELOPING.adoc[Developing] for details about our development process and how to build and test the project.

== Copyright and License

Copyright (c) 2022 D4L data4life gGmbH / All rights reserved.

Please refer to our link:LICENSE[License] for further details.


CHANGELOG.adoc
= Changelog
:link-repository: https://github.com/d4l-data4life/hc-crypto-sdk-kmp
:doctype: article
:toc: macro
:toclevels: 1
:toc-title:
:icons: font
:imagesdir: assets/images
ifdef::env-github[]
:warning-caption: :warning:
:caution-caption: :fire:
:important-caption: :exclamation:
:note-caption: :paperclip:
:tip-caption: :bulb:
endif::[]

All notable changes to this project will be documented in this file.

The format is based on http://keepachangelog.com/en/1.0.0/[Keep a Changelog]
and this project adheres to http://semver.org/spec/v2.0.0.html[Semantic Versioning].

Template could be found here: link:https://github.com/d4l-data4life/hc-readme-template/blob/main/TEMPLATE_CHANGELOG.adoc[CHANGELOG]

[discrete]
==== Versions

toc::[]


== link:{link-repository}/releases/latest[Unreleased]

See link:{link-repository}/compare/v1.16.1\...main[changeset]

=== Added

=== Changed

=== Deprecated

=== Removed

=== Fixed

=== Security

=== Bumped

=== Migration

== link:{link-repository}/releases/v1.16.1[1.16.1]

See link:{link-repository}/compare/v1.16.0\...v1.16.1[changeset]

=== Fixed

* KAPT stopped working for Android

== link:{link-repository}/releases/tag/v1.16.0[1.16.0]

See link:{link-repository}/compare/v1.15.0\...v1.16.0[changeset]

=== Bumped

* Gradle 7.4.2 -> 7.5.1
* Android Gradle Plugin 7.1.3 -> 7.2.2
* *_BREAKING_* Kotlin 1.6.10 -> 1.7.10
* *_BREAKING_* Kotlin Coroutines 1.6.0 -> 1.6.3
* *_BREAKING_* Kotlin Serialization 1.3.2 -> 1.3.3
* *_BREAKING_* Kotlin DateTime 0.3.2 -> 0.4.0
* Android Ktx 1.7.0 -> 1.8.0
* Android AppCompat 1.4.1 -> 1.5.0
* Android ConstraintLayout 2.1.3 -> 2.1.4.
* Android Material 1.5.0 -> 1.6.1
* Android Target SDK 31 -> 33
* HC Util 1.13.0 -> 1.14.0

== link:{link-repository}/releases/tag/v1.15.0[1.15.0]

See link:{link-repository}/compare/v1.14.0\...v1.15.0[changeset]

=== Removed

* HC Result

=== Bumped

* Gradle 7.2 -> 7.4.2
* Android Gradle Plugin 7.0.2 -> 7.1.3
* *_BREAKING_* Kotlin 1.5.21 -> 1.6.10
* *_BREAKING_* Kotlin Coroutines 1.5.1 -> 1.6.0
* *_BREAKING_* Kotlin Serialization 1.2.2 -> 1.3.2
* *_BREAKING_* Kotlin DateTime 0.2.1 -> 0.3.2
* HC Util 1.11.0 -> 1.13.0
* Moshi 1.12.0 -> 1.13.0
* BouncyCastle 1.64 -> 1.71
* Desugar 1.0.4 -> 1.1.5
* Android Ktx 1.3.2 -> 1.7.0
* Android AppCompat 1.3.0 -> 1.4.1
* Android Browser 1.2.0 -> 1.4.0
* Android ConstraintLayout 2.0.1 -> 2.1.3
* Android Material 1.4.0 -> 1.5.0
* Android Target SDK 30 -> 31

== link:{link-repository}/releases/tag/v1.14.0[1.14.0]

See link:{link-repository}/compare/v1.13.2\...v1.14.0[changeset]

=== Added

* Add Serializable that got removed from HC Util SDK
* Add HC Result SDK 0.3.0

=== Bumped

* HC Util SDK 1.7.0 -> 1.11.0
* Gradle 6.9 -> 7.2
* *_BREAKING_* Kotlin 1.4.32 -> 1.5.21
* *_BREAKING_* Android Gradle Plugin 4.2.2 -> 7.0.2
* *_BREAKING_* Coroutines 1.4.3 -> 1.5.1
* Kotlin Serialization 1.0.1 -> 1.2.2
* Kotlin date-time 0.1.1 -> 0.2.1
* Android Material Components 1.3.0 -> 1.4.0
* Robolectric 4.5.1 -> 4.6.1

== link:{link-repository}/releases/tag/v1.13.2[1.13.2]

Initial release taken out of link:https://github.com/d4l-data4life/hc-sdk-kmp/[HC SDK KMP].


buildSrc/src/main/kotlin/care/data4life/sdk/crypto/LibraryConfig.kt
/*
 * Copyright (c) 2021 D4L data4life gGmbH / All rights reserved.
 *
 * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
 * including any intellectual property rights that subsist in the SDK.
 *
 * The SDK and its documentation may be accessed and used for viewing/review purposes only.
 * Any usage of the SDK for other purposes, including usage for the development of
 * applications/third-party applications shall require the conclusion of a license agreement
 * between you and D4L.
 *
 * If you are interested in licensing the SDK for your own applications/third-party
 * applications and/or if you’d like to contribute to the development of the SDK, please
 * contact D4L by email to help@data4life.care.
 */

package care.data4life.sdk.crypto

object LibraryConfig {

    val publishConfig = PublishConfig
    val android = AndroidLibraryConfig

    const val group = "care.data4life.sdk.crypto"
    const val name = "hc-crypto-sdk-kmp"

    const val githubOwner = "d4l-data4life"
    const val githubRepository = "hc-crypto-sdk-kmp"

    object PublishConfig {
        const val groupId = "care.data4life.hc-crypto-sdk-kmp"

        const val description = "Implementation of the D4L Crypto Protocol"
        const val year = "2021"

        // URL
        const val host = "github.com"
        const val path = "$githubOwner/$githubRepository"

        const val url = "https://$host/$path"

        // DEVELOPER
        const val developerId = "d4l-data4life"
        const val developerName = "D4L data4life gGmbH"
        const val developerEmail = "mobile@data4life.care"

        // LICENSE
        const val licenseName = "Private"
        const val licenseUrl = "$url/blob/main/LICENSE"
        const val licenseDistribution = "repo"

        // SCM
        const val scmUrl = "git://$host/$path.git"
        const val scmConnection = "scm:$scmUrl"
        const val scmDeveloperConnection = scmConnection
    }

    object AndroidLibraryConfig {
        const val minSdkVersion = 23
        const val compileSdkVersion = 30
        const val targetSdkVersion = 30

        const val resourcePrefix = "hc_crypto_"
    }
}
